### PR TITLE
Creates a weekly worklfow to request a build on all active flavors

### PR DESCRIPTION
**Pull request by addyess**
### Overview

A new workflow to rebuild Release Branches and Main weekly

### Details
* a new workflow calling request_builds.py script
* move many utility methods into a new util structure for reuse
    * launchpad (lp)
    * git methods (repo)
    * snapstore
    * k8s specific utils

### CODE REVIEW

File: scripts/request_builds.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        "--branches", nargs="*", type=str, help="Specific branches to confirm"
+    )
+    args = util.setup_arguments(arg_parser)
+    branches = args.branches
+
+    if not branches:
+        all_branches = repo.ls_branches(util.SNAP_REPO)
+        branches = [b for b in all_branches if util.TIP_BRANCH.match(b)]
+        LOG.info("No branches specified, checking '%s'", ", ".join(branches))
+    rebuild_branches(filter_branches(branches), args)
```
Comment (berkayoz) [inline scripts/request_builds.py:89]:
```suggestion
    branches = args.branches

    if not branches:
        all_branches = repo.ls_branches(util.SNAP_REPO)
        LOG.info("No branches specified, checking '%s'", ", ".join(branches))
    rebuild_branches(filter_branches(branches), args)
```

Is this required, wouldn't this skip the `if not util.TIP_BRANCH.match(branch):` in `fillter_branches`?

**Review by berkayoz — COMMENTED** [review]
Great work, one small comment.

File: scripts/request_builds.py (hunk 1 above)
Comment (addyess) [reply scripts/request_builds.py]: Thanks @berkayoz . that's a super nice spot.  It helped me find a bug i didn't know was lurking.

**Comment by addyess** [comment]
sample output from running locally with `dry-run`

```
      request_builds 2024-09-25 08:39:00,283     INFO - No branches specified, checking all branches
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1284/join-test' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1284/node-status-test' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1400-Add-howto-for-EPA-docs' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1468/duplicate-remove' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1513/capi-refresh-certs' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1517/poc-refactor-network-tests' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1586' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1598/placeholder-migration' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-1658/cilium-local-apiserver' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'KU-308/core24-upgrade' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'archive/release-1.30-moonray' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'autoupdate/moonray' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'autoupdate/release-1.30-strict' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'autoupdate/release-1.31-moonray' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'autoupdate/release-1.31-strict' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'autoupdate/strict' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'berkayoz/cilium-kube-proxy-replacement' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'bschimke95/fix-dqlite-migrations' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'bschimke95/make-token-errors-more-precise' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'dev-1.29.6' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'dev/etcd' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'dev/metallb' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'eac/snapd-config' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,823  WARNING - Skipping branch 'eaudetcobello-patch-1' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,824  WARNING - Skipping branch 'fix-path-capi-tutorial-getting-started-doc' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:02,824  WARNING - Skipping branch 'investigate-CAPI-failed-node-removal' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:03,029     INFO - Cloning tip branch main
           util.repo 2024-09-25 08:39:03,030     INFO - Cloning https://github.com/canonical/k8s-snap.git/ @ main (shallow=True)
      request_builds 2024-09-25 08:39:04,699     INFO -   Kubernetes version detected v1.31.1
      request_builds 2024-09-25 08:39:04,701     INFO -   Searching for recipe k8s-snap-tip-strict
      request_builds 2024-09-25 08:39:05,126     INFO -   Requesting build for k8s-snap-tip-strict (dry-run)
      request_builds 2024-09-25 08:39:05,126     INFO -   Searching for recipe k8s-snap-tip-moonray
      request_builds 2024-09-25 08:39:05,462     INFO -   Requesting build for k8s-snap-tip-moonray (dry-run)
      request_builds 2024-09-25 08:39:05,462     INFO -   Searching for recipe k8s-snap-tip-classic
      request_builds 2024-09-25 08:39:05,842     INFO -   Requesting build for k8s-snap-tip-classic (dry-run)
      request_builds 2024-09-25 08:39:05,852  WARNING - Skipping branch 'moonray/main' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:06,066     INFO - Cloning tip branch release-1.30
           util.repo 2024-09-25 08:39:06,066     INFO - Cloning https://github.com/canonical/k8s-snap.git/ @ release-1.30 (shallow=True)
      request_builds 2024-09-25 08:39:07,243     INFO -   Kubernetes version detected v1.30.5
      request_builds 2024-09-25 08:39:07,245     INFO -   Searching for recipe k8s-snap-1.30-strict
      request_builds 2024-09-25 08:39:07,522     INFO -   Requesting build for k8s-snap-1.30-strict (dry-run)
      request_builds 2024-09-25 08:39:07,522     INFO -   Searching for recipe k8s-snap-1.30-classic
      request_builds 2024-09-25 08:39:07,811     INFO -   Requesting build for k8s-snap-1.30-classic (dry-run)
      request_builds 2024-09-25 08:39:08,073     INFO - Cloning tip branch release-1.31
           util.repo 2024-09-25 08:39:08,073     INFO - Cloning https://github.com/canonical/k8s-snap.git/ @ release-1.31 (shallow=True)
      request_builds 2024-09-25 08:39:09,726     INFO -   Kubernetes version detected v1.31.1
      request_builds 2024-09-25 08:39:09,728     INFO -   Searching for recipe k8s-snap-1.31-strict
      request_builds 2024-09-25 08:39:09,983     INFO -   Requesting build for k8s-snap-1.31-strict (dry-run)
      request_builds 2024-09-25 08:39:09,983     INFO -   Searching for recipe k8s-snap-1.31-moonray
      request_builds 2024-09-25 08:39:10,345     INFO -   Requesting build for k8s-snap-1.31-moonray (dry-run)
      request_builds 2024-09-25 08:39:10,345     INFO -   Searching for recipe k8s-snap-1.31-classic
      request_builds 2024-09-25 08:39:10,708     INFO -   Requesting build for k8s-snap-1.31-classic (dry-run)
      request_builds 2024-09-25 08:39:10,718  WARNING - Skipping branch 'runc-patch-v1.1.13' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
      request_builds 2024-09-25 08:39:10,718  WARNING - Skipping branch 'strict/main' - not a tip branch r/^(?:main)|^(?:release-\d+\.\d+)$/
```

**Review by bschimke95 — APPROVED** [review]
LGTM, minor nits but looks great - I appreciate the util module being split up.

File: scripts/util/snapstore.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+            return json.loads(response.read().decode())
+    except HTTPError as e:
+        LOG.exception("HTTPError ({%s}): {%s} {%s}", req.full_url, e.code, e.reason)
+        raise
+    except URLError as e:
+        LOG.exception("URLError ({%s}): {%s}", req.full_url, e.reason)
+        raise
+
+
+def ensure_track(snap_name: str, track_name: str):
```
[the remark this answers is not in the document]
Comment (addyess) [reply scripts/util/snapstore.py:29]: yes -- held for future PR

File: scripts/util/util.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import semver
+
+LOG = logging.getLogger(__name__)
+SNAP_NAME: str = "k8s"
+SNAP_REPO: str = "https://github.com/canonical/k8s-snap.git/"
+LP_OWNER: str = "containers"
+TIP_BRANCH = re.compile(r"^(?:main)|^(?:release-\d+\.\d+)$")
+
+
+def flavors(dir: str) -> list[str]:
```
[the remark this answers is not in the document]
Comment (addyess) [reply scripts/util/util.py:16]: can do

File: .github/workflows/rebuild-release-branch.yaml [hunk 4]
```diff
@@ -0,0 +1,39 @@
+name: Rebuild Release Branch
+
+on:
+  workflow_dispatch:
+    inputs:
+      branches:
+        type: string
+        required: false
+        default: ""  # defaults to all matching branches (main and release branches)
```
[the remark this answers is not in the document]
Comment (addyess) [reply .github/workflows/rebuild-release-branch.yaml:9]: There's another job for ensuring (and failing) if there are recipes for each existing branch. 

There's another card to make sure that on a new upstream release we create branches, sync branches, create recipes, and create tracks

**Pull request merged**